bitcoin/src/psbt/mod.rs

```diff
@@ -446,6 +559,64 @@ impl Psbt {
         }
     }
 
+    /// Returns the sighash message to sign an SCHNORR input along with the sighash type.
+    ///
+    /// Uses the [`TapSighashType`] from this input if one is specified. If no sighash type is
+    /// specified uses [`TapSighashType::Default`].
+    fn sighash_taproot<T: Borrow<Transaction>>(
+        &self,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+        leaf_hash: Option<TapLeafHash>
+    ) -> Result<(Message, TapSighashType), SignError> {
+        use OutputType::*;
+
+        if self.signing_algorithm(input_index)? != SigningAlgorithm::Schnorr {
+            return Err(SignError::WrongSigningAlgorithm);
+        }
+
+        let input = self.checked_input(input_index)?;
+
+        match self.output_type(input_index)? {
+            Tr => {
+                let hash_ty = input
+                    .sighash_type
+                    .unwrap_or_else(|| TapSighashType::Default.into())
+                    .taproot_hash_ty()
+                    .map_err(|_| SignError::InvalidSighashType)?;
+
+                let witness_utxos = (0..self.inputs.len())
+                    .map(|i|  self.spend_utxo(i).ok())
+                    .collect::<Vec<_>>();
+                let mut all_witness_utxos = vec![];
+
+                let is_anyone_can_pay = PsbtSighashType::from(hash_ty).to_u32() & 0x80 != 0;
+
+                let prev_outs = if is_anyone_can_pay {
+                    Prevouts::One(
+                        input_index,
+                        witness_utxos[input_index].ok_or(SignError::MissingSpendUtxo)?,
+                    )
+                } else if witness_utxos.iter().all(Option::is_some) {
+                    all_witness_utxos.extend(witness_utxos.iter().filter_map(|x| x.as_ref()));
```

Comment:
> Same comment as for `witness_utxos`. It would be cool if we could do this without allocating a new vector but I couldn't work out how, maybe one of the other lads will be able to.

If I haven't misunderstood you, let's see how this new implementation works.